### ISSUE TITLE
[minor] scatter_ doesn't support empty index vector

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#2027]
Nop operations on empty tensors are sometimes convenient to save boiler-plate code on corner cases. PyTorch accepts many operations on empty tensors without a problem:
```python
a = torch.Tensor()
b = (a - 1) * 5
```

but fails for `scatter_`:
```python
torch.zeros(10).scatter_(0, torch.LongTensor(), 0.0)
```

**Comment by elanmart** [comment]
Hi, 

is this indeed an undesried behaviour? Looking at the code in [torch/lib/TH/generic/THTensorMath.c](https://github.com/pytorch/pytorch/blob/f608208a8051a8708e71291e84768bf8be720a1a/torch/lib/TH/generic/THTensorMath.c) it seems like most of indexing-related functions will raise an error if called with 
empty index (e.g. `gather`, `index_select`). 

If this is not desired, how could this be handled nicely? I imagine simply returning
from a function if we detect an empty index is not a cool solution?

**Comment by vadimkantorov** [comment]
To me, that's a fine solution. `scatter_` does the operation in-place, so returning the original tensor is what the user expects anyway (copying version `Variable.scatter(...)` should probably return a copy of the original variable).

**Comment by vadimkantorov** [comment]
On master the error message is quite strange:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: invalid argument 3: Index tensor must have same dimensions as output tensor at .../pytorch/aten/src/TH/generic/THTensorMath.c:661
```

**Comment by ppwwyyxx** [comment]
A related issue for `index_select` has been fixed: #3416
